app/src/main/java/app/revanced/integrations/shared/checks/CheckEnvironmentPatch.java

```diff
@@ -0,0 +1,307 @@
+                    equalsHash(Build.BRAND, PATCH_BRAND) &&
+                    equalsHash(Build.CPU_ABI, PATCH_CPU_ABI) &&
+                    equalsHash(Build.CPU_ABI2, PATCH_CPU_ABI2) &&
+                    equalsHash(Build.DEVICE, PATCH_DEVICE) &&
+                    equalsHash(Build.DISPLAY, PATCH_DISPLAY) &&
+                    equalsHash(Build.FINGERPRINT, PATCH_FINGERPRINT) &&
+                    equalsHash(Build.HARDWARE, PATCH_HARDWARE) &&
+                    equalsHash(Build.HOST, PATCH_HOST) &&
+                    equalsHash(Build.ID, PATCH_ID) &&
+                    equalsHash(Build.MANUFACTURER, PATCH_MANUFACTURER) &&
+                    equalsHash(Build.MODEL, PATCH_MODEL) &&
+                    equalsHash(Build.ODM_SKU, PATCH_ODM_SKU) &&
+                    equalsHash(Build.PRODUCT, PATCH_PRODUCT) &&
+                    equalsHash(Build.RADIO, PATCH_RADIO) &&
+                    equalsHash(Build.SERIAL, PATCH_SERIAL) &&
+                    equalsHash(Build.SKU, PATCH_SKU) &&
+                    equalsHash(Build.SOC_MANUFACTURER, PATCH_SOC_MANUFACTURER) &&
+                    equalsHash(Build.SOC_MODEL, PATCH_SOC_MODEL) &&
+                    equalsHash(Build.TAGS, PATCH_TAGS) &&
+                    equalsHash(Build.TYPE, PATCH_TYPE) &&
+                    equalsHash(Build.USER, PATCH_USER);
+
+            if (passed) {
+                Logger.printDebug(() -> "Build properties match current device");
+            } else {
+                Logger.printDebug(() -> "Build properties do not match current device");
+            }
+
+            return passed;
+        }
+    };
+
+    /**
+     * Check if the app was installed within the last 30 minutes after being patched.
+     * <br>
+     * If the app was installed within the last 30 minutes, it is likely, the app was patched by the user.
+     * <br>
+     * If the app was installed at a later time, it is likely, the app was downloaded pre-patched, the user
+     * waited too long to install the app or the patch time is too long ago.
+     */
+    private static final Check isNearPatchTimeCheck = new Check(
+            "revanced_check_environment_not_near_patch_time"
+    ) {
+        @Override
+        protected boolean run() {
+            final var duration = System.currentTimeMillis() - PATCH_TIME;
+
+            Logger.printDebug(() -> "Installed in " + duration / 1000 + " seconds after patch");
+
+            return duration < 1000 * 60 * 30; // 1000 ms * 60 s * 30 min;
+        }
+    };
+
+    /**
+     * Check if the IP address is the same as during the patch.
+     * <br>
+     * If the IP address is the same as during the patch, it is likely, the app was just patched by the user.
+     * <br>
+     * If the IP address is different, the app was likely downloaded pre-patched or the patch time is too long ago.
+     */
+    private static final Check hasPatchTimePublicIPCheck = new Check(
```

Comment:
The only issue I have with this check is that it relies on external services to pass. The check in itself just relies on the hash of the last couple characters of the IP, so the hash can't be brute forced either to resolve the IP. In the case of a CLI installation the IP check only works on devices behind NAT. In the case of IPv6 we could simply check for the IPv6 prefix and therefore get rid of the external service call entirely. So an additional check in this one, if the user uses IPv6 should prevent an external call for those who use IPv6. For IPv4 the subnet could be checked or the local device where CLI was used, pinged, but both could be a bit unreliable.

Sincerely,
oSumAtrIX

Reply:
> Although, if installing thru CLI it will still pass since the 'patch time' check will pass.
> 
> Should this check be kept?

You may still install it a bit later than 30 min.

> Sometimes this ip hash check is failing, and I'm not sure why.

If you have both IPv4 and IPv6, check if the sites randomly switch between them, could be the reason. 

> I think the ip check failed because of whitespace or other issues (that can be fixed). But just for privacy reasons I think this check should go, because the other checks of patch time and installer source will work for CLI installs without needing to check or rely on a third party service.

The only check that would be left for CLI is the time check, all other won't work because Android doesn't allow seing other apps and if you are using CLI probably you are missing ReVanced Manager anyways. The IP check is only ran after all the other checks so it is very unlikely for normal users to encounter. Users of rip off sites would be more likely affected. Currently only if all checks fail, the warning is shown, but perhaps a specific combination of checks could result in a warning, skipping the IP check. Additionally, next to the external IPv4 check other checks can be made such as the local IPv6 prefix or subnet and similar making an external call even less likely.

Reply:
Sounds like it could work. I guess we can test this approach